Place small snow slabs during snowy weather

RatMoleRat opened the pull request:
This implements the idea in issue #2406 to make snowy weather deposits snow slabs (`engine:eighthBlock` is the shape I used for the slabs). There seems to be a fair hit to performance, so any suggestions to improve the efficiency would be welcome. 

Outstanding before merging:
 - [X] Add logic to make the snow blocks melt during sunny weather (they currently stay in place until removed by the player)

Cervator commented:
Can confirm this works! And is beautiful to see :-)

As for performance I've noticed it from the particles too, probably from inefficiencies in that system and the sheer amount of particles. I'd doubt the new snow placing system really adds much. I do have some architectural recommendations though

Primarily one big problem right now is that the two main systems are plain `@RegisterSystem` meaning they'll in theory run on _both_ clients and servers (the authority) - even on a headless server. For the particle effects at least that's just silly. It may also mean in a single player game some things are running twice (bad memory there so not sure). And naturally it means havoc when actually trying multiplayer - specifically a solo-client trying to connect will crash.

So my suggestion is to rework the systems into this set of things:

* An authority-only system that just determines the weather and sends begin/end events to all connected clients so they can react
* A client-only system that runs the particle effects and nothing else. Listens to the broadcast events to know when to start and end particle effects
* An authority-only system that makes snow pile up. This _could_ be in the other authority system but keeping them separate is nice isolation and would allow for the advanced weather effects to be made optional later (could even be in a different module if needed)

Note how the authority system making snow pile up would then not be able to rely on the local player _at all_ - you'd want the logic to run solely on the server. But then where do we pile up the snow? That's more of an open-ended design design question. Should weather be global or localized? Should biomes get involved? How about temperature maps and maybe those melt snow rather than sun?

I'd say for now have the authority retrieve the list of online players then process all the in-range chunks (once, not per player - just sum up all the affected chunks then process). So each player gets their own little snow cloud - much like how at present the particles are only visible around a player, but in reality the two are complete separate: we fake display weather around players locally only + have the server simulate weather where they happen to be.

One other simple suggestion: Favor a dependency on Core to get the snow block rather than re-add it here. That may seem like a a heavy dependency just to get a block and it is - a longer term idea would be to switch that dependency to a base asset pack later split out from Core. Or a secondary module doing snow piles later could have that dependency. Easy enough to include right now for testing though 👍 

(Fun fact: optional dependencies could mean the snow piling system would depend on Core being present and simply not run if it isn't there and snow as a block isn't available - another benefit of splitting out the system)

Other fun suggestions: Make the snow pile up in increasing depths ... make it ice over water etc too :-)

Nicely done!

RatMoleRat commented:
I moved around the systems as was suggested, and added a dependency on Core (though I found the Core snowball looks odd when used in the `eighthblock` shape - not sure why). A bit of testing in multiplayer indicates that both players can connect. I couldn't test much further in multiplayer since I was running both game instances on the same computer and was only getting a couple FPS.

jellysnake commented on src/main/java/org/terasology/weatherManager/systems/BlockPlacingWeatherSystem.java:
```diff
@@ -0,0 +1,137 @@
+     */
+    @ReceiveEvent
+    public void onPlaceEvent(PeriodicActionTriggeredEvent event, EntityRef worldEntity) {
+        Iterator<Client> players = networkSystem.getPlayers().iterator();
```
So this is a really expensive method actually, because both the get players and the iteration will happen for _every_ periodic event.
I know you filter later, but i'd like to see the filtering as the first line in this method.

jellysnake commented on src/main/java/org/terasology/weatherManager/systems/BlockPlacingWeatherSystem.java:
```diff
@@ -0,0 +1,137 @@
+    @ReceiveEvent
+    public void onPlaceEvent(PeriodicActionTriggeredEvent event, EntityRef worldEntity) {
+        Iterator<Client> players = networkSystem.getPlayers().iterator();
+        while (players.hasNext()) {
```
Any particular reason you're using this over
```
for(Client currentPlayer : networkSystem.getPlayers()) {
    // Do Stuff
}
```

jellysnake commented on src/main/java/org/terasology/weatherManager/systems/BlockPlacingWeatherSystem.java:
```diff
@@ -0,0 +1,137 @@
+    }
+
+    private void placeSnow(Vector3i playerPos) {
+        Random rand = new Random();
```
Prefer using `FastRandom` over the default random

jellysnake commented on src/main/java/org/terasology/weatherManager/systems/BlockPlacingWeatherSystem.java:
```diff
@@ -0,0 +1,137 @@
+            } else if (current.equals(air)) {
+                currentY--;
+                lastGround = false;
+            } else if (current.isWaving() || !current.isAttachmentAllowed()) {
```
Prefer checking solid not waving imo

jellysnake commented on src/main/java/org/terasology/weatherManager/systems/BlockPlacingWeatherSystem.java:
```diff
@@ -0,0 +1,137 @@
+                currentY--;
+                lastGround = false;
+            } else if (current.isWaving() || !current.isAttachmentAllowed()) {
+                placed = true; //break out to avoid placing snow on blocks like grass
```
Why don't you just actually `break;` rather than this?

jellysnake commented on src/main/java/org/terasology/weatherManager/systems/WeatherManagerSystem.java:
```diff
@@ -142,9 +146,9 @@ private void changeWeather(ConditionAndDuration conditionAndDuration) {
+        triggerEvents();
 
-        weatherEntity.send(new StartWeatherEvent());
+        logger.info("WEATHER CHANGED: " + current.condition + "(" + current.duration + ")");
```
No logger statements in final PR's please :)

jellysnake commented:
I feel like this is ready to merge.
The multiplayer hardening and future refactoring is probably best suited to a new PR. This one already ballooned from adding snow deposits to changing a swath of the system @Cervator  if you are happy with it i'll merge tomorrow morning
